packages/iov-bns/src/types.ts

```diff
@@ -23,6 +23,89 @@ export interface CashConfiguration {
   readonly minimalFee: Amount;
 }
 
+// Governance
+
+export interface Elector {
+  readonly address: Address;
+  /** The voting weight of this elector. Max value is 65535 (2^16-1). */
+  readonly weight: number;
+}
+
+export interface Electorate {
+  readonly id: number;
+  readonly version: number;
+  readonly admin: Address;
+  readonly title: string;
+  /** An unordered map from address to elector */
+  readonly electors: { readonly [index: string]: Elector };
+  /** Sum of all electors' weights */
+  readonly totalWeight: number;
+}
+
+export interface Fraction {
+  readonly numerator: number;
+  readonly denominator: number;
+}
+
+export interface ElectionRule {
+  readonly id: Uint8Array;
+  readonly version: number;
+  readonly admin: Address;
+  /** The electorate used by this (ID without version, as changing electorate changes the rule) */
+  readonly electorateId: number;
+  readonly title: string;
+  /** Voting period in seconds */
+  readonly votingPeriod: number;
+  readonly threshold: Fraction;
+  readonly quorum: Fraction | null;
+}
+
+export interface VersionedId {
+  readonly id: Uint8Array;
+  readonly version: number;
+}
+
+export enum ProposalExecutorResult {
+  NotRun,
+  Succeeded,
+  Failed,
+}
+
+export enum ProposalResult {
+  Undefined,
+  Accepted,
+  Rejected,
+}
+
+export enum ProposalStatus {
+  Submitted,
+  Closed,
+  Withdrawn,
+}
+
+/** Union type for possible options */
+export type ProposalOption = string;
+
+export interface Proposal {
+  readonly title: string;
+  /** The transaction to be executed when the proposal is accepted */
+  readonly option: ProposalOption;
+  readonly description: string;
+  readonly electionRule: VersionedId;
+  readonly electorate: VersionedId;
```

Comment:
Do we need the version here? Are we going to have forking electorates on a single chain?

Reply:
Assume you have an electorate "tech committee". People can leave and join this commitee in a slow and trusted fashion. Whenever this committee changes, the `version` from the `(id, version)` pair is bumped, but the `id` is not. This allows you to reference the latest version of the tech comittee when creating a proposal. As soon as the proposal is created on chain (i.e. the create transaction is executed), the version is fixed. This implicitely fixes all members and their weights.

Reply:
Right, so as soon as the committee changes we need new versions of all the rules we want to continue, right? And is there going to be some mechanism on the backend preventing proposals referencing older versions of an electorate?

Reply:
1. A proposal is created without the need to know the latest version (which is considered a feature). In [CreateProposalMsg]( https://htmlpreview.github.io/?https://github.com/iov-one/weave/blob/v0.16.0/docs/proto/index.html#gov.CreateProposalMsg), we only set the `id` of `(id, version)` for both ElectionRule and implicitly Electorate.
2. Due to 1., at proposal creation time the blockchain picks the latest version. As soon as a poposal is created, the version is fixed.